MovieMVP/MovieMVP/MoviesTableViewModule/Presenter/MoviesPresenter.swift

```diff
@@ -18,26 +20,59 @@ final class MoviesPresenter: MoviesPresenterProtocol {
     init(
         view: MoviesProtocol,
         networkService: NetworkServiceProtocol,
+        realmService: RealmServiceProtocol,
+        imageService: ImageCacheServiceProtocol,
         router: RouterProtocol
     ) {
         self.view = view
         self.networkService = networkService
         self.router = router
+        self.realmService = realmService
+        self.imageService = imageService
     }
 
     // MARK: - Methods
 
     func fetchMovies(category: Category) {
-        networkService.fetchMovies(category: category) { [weak self] result in
+        networkService?.fetchMovies(category: category) { [weak self] result in
             guard let self = self else { return }
             switch result {
             case let .success(movies):
+                movies.forEach { movie in
+                    movie.movieCollection = category.categoryString
+                }
                 self.movies = movies
                 self.view?.success()
             case let .failure(error):
+
                 self.view?.failure(error: error)
             }
         }
+        realmService.saveData(movies ?? [])
+    }
+
+    private func loadData(category: Category) {
+        guard let items = realmService.readMovieData(type: Movie.self, movieCollection: category.categoryString)
+        else { return }
+        if !items.isEmpty {
+            movies = Array(items)
+            view?.success()
+        } else {
+            fetchMovies(category: category)
+        }
+    }
+
+    func fetchMoviesType(index: Int) {
+        switch index {
+        case 0:
+            loadData(category: .popular)
+        case 1:
+            loadData(category: .top)
+        case 2:
+            loadData(category: .upcoming)
+        default:
+            break
+        }
     }
 
     func giveMovieID(id: Int) {
```

Comment:
```suggestion
    func giveMovieID(id: Int) {
```
тоже поднять выше нужно